---
title: "SIG04 10x Seurat Processing"
author: "Eric Y. Wang"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: true
    toc_float: true
  github_document:
    toc: true
    html_preview: false
---

```{r setup, include=FALSE}
library(tidyverse)
library(Seurat)
library(hdf5r)
library(ggplot2)
library(clustree)
library(cowplot)
library(patchwork)
library(ggpointdensity)
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
source("functions/plotting_fxns.R")
source("functions/scRNA_seq_analysis_functions.R")
theme_set(theme_Publication())
```

## [Import data]{.underline}

```{r}
dataRaw1 <- Read10X_h5("C:/Users/Eric/My Drive/Lab/datasets/EYW/SIG04_10x_240816/cellranger_outs/SIG04_lane1_GEX_oBC/filtered_feature_bc_matrix.h5")
dataRaw2 <- Read10X_h5("C:/Users/Eric/My Drive/Lab/datasets/EYW/SIG04_10x_240816/cellranger_outs/SIG04_lane2_GEX_oBC/filtered_feature_bc_matrix.h5")

names(dataRaw1)
names(dataRaw2)
```

```{r}
subset_joint <- function(x){
  # identify joint cells between BC and GEX dataset
  jointCells <- intersect(colnames(x$`Gene Expression`),
                          colnames(x$`CRISPR Guide Capture`))
  
  # subset each dataset to include joint cells
  data <- vector(mode = "list")
  data[["GEX"]] <- x$`Gene Expression`[,jointCells]
  data[["oBC"]] <- x$`CRISPR Guide Capture`[,jointCells]
  
  return(data)
}

dataRaw1 <- subset_joint(dataRaw1)
dataRaw2 <- subset_joint(dataRaw2)
```


### Filter Low Genes

Remove genes that are expressed by less than 1 in 1000 cells (assuming 30% dropout
rate). Set this level based on the smallest size cluster that you want to be
detectable in the final dataset.

```{r}
dataMergeGEX <- cbind(dataRaw1$GEX,dataRaw2$GEX)

# create vector of genes to keep
keepGenes <- Matrix::rowSums(dataMergeGEX > 0) >= ncol(dataMergeGEX)*(1/1000)*0.3

# remove genes from gex matrices
dataRaw1$GEX <- dataRaw1$GEX[keepGenes,]
dataRaw2$GEX <- dataRaw2$GEX[keepGenes,]

rm(dataMergeGEX)
```

### Create Seurat

```{r}
create_seurat <- function(x){
  # load GEX counts and add celltag counts into seurat object
  data <- CreateSeuratObject(counts = x$GEX, project = "SIG04")
  data[["oBC"]] <- CreateAssayObject(counts = x$oBC)

  return(data)
}

dataSeurat <- vector(mode = "list")
dataSeurat[["lane1"]] <- create_seurat(dataRaw1)
dataSeurat[["lane2"]] <- create_seurat(dataRaw2)

rm(dataRaw1,dataRaw2)
```

## [Cell QC Analysis]{.underline}

### Calculate QC metrics

```{r}
add_qc_metrics <- function(x){
  DefaultAssay(x) <- "RNA"
  # calculate percent mitochondrial reads and ratio 
  x[["percent.mt"]] <- PercentageFeatureSet(x, pattern = "^mt-")
  x[["percent.ribo"]] <- PercentageFeatureSet(x, pattern = "^Rpl|^Rps")
  x[["mitoRatio"]] <- x@meta.data$percent.mt / 100
  x[["log10GenesPerUMI"]] <- log10(x$nFeature_RNA)/log10(x$nCount_RNA)
  x[["log10nFeature_RNA"]] <- log10(x$nFeature_RNA)
  x[["log10nCount_RNA"]] <- log10(x$nCount_RNA)
  
  return(x)
}

dataSeurat <- lapply(dataSeurat, function(x) add_qc_metrics(x))

metadata <- bind_rows(dataSeurat[[1]]@meta.data,dataSeurat[[2]]@meta.data) %>%
  mutate(orig.ident = c(rep("lane1",ncol(dataSeurat[[1]])),rep("lane2",ncol(dataSeurat[[2]]))))
```

### Visualize QC distributions

```{r, fig.width=15, fig.height=7}
# enter QC values
qcValues <- c(nFeature_RNA = 1000, nCount_RNA = 2000, percent.mt = 5)

# Visualize the number UMIs/transcripts per cell
p1 <- metadata %>%
  	ggplot(aes(color=orig.ident, x=nCount_RNA, fill= orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	ylab("Cell density") +
  	geom_vline(xintercept = qcValues[2]) +
    NoLegend()

# Visualize the distribution of genes detected per cell via histogram
p2 <- metadata %>% 
  	ggplot(aes(color=orig.ident, x=nFeature_RNA, fill= orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	geom_vline(xintercept = qcValues[1]) +
    NoLegend()

# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
p5 <- metadata %>% 
  	ggplot(aes(x=nCount_RNA, y=nFeature_RNA, color=percent.mt)) + 
  	geom_point() + 
  	scale_colour_gradient(low = "gray90", high = "black") +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	geom_vline(xintercept = qcValues[2]) +
  	geom_hline(yintercept = qcValues[1])


# Visualize the distribution of mitochondrial gene expression detected per cell
p3 <- metadata %>% 
  	ggplot(aes(color=orig.ident, x=percent.mt, fill=orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	geom_vline(xintercept = qcValues[3]) +
    xlim(0,10) +
    NoLegend()

# Visualize the overall complexity of the gene expression by visualizing the genes detected per UMI
p4 <- metadata %>%
  	ggplot(aes(x=log10GenesPerUMI, color = orig.ident, fill=orig.ident)) +
  	geom_density(alpha = 0.2) +
  	geom_vline(xintercept = 0.8)

# Visualize the correlation between genes detected and number of UMIs with mt filter
p6 <- metadata %>%
    filter(percent.mt < qcValues[3]) %>%
  	ggplot(aes(x=nCount_RNA, y=nFeature_RNA)) + 
    geom_pointdensity() +
    scale_color_viridis_c() +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	geom_vline(xintercept = qcValues[2]) +
  	geom_hline(yintercept = qcValues[1])

plot_grid(p1,p2,p3,p4,p5,p6, ncol = 3)
```

```{r, fig.height=5, fig.width=6}
metadata %>%
  	ggplot(aes(color=orig.ident, x=log10(nCount_oBC+1), fill= orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	ylab("Cell density")
```
```{r, fig.height=5, fig.width=12}
metadata %>%
  	ggplot(aes(x=log10(nCount_oBC+1), y=log10(nCount_RNA+1))) + 
    facet_wrap(~orig.ident) +
    geom_pointdensity() +
    scale_color_viridis_c()
```

It's weird that lane 2 which consists of sorted cells has such a skewed distribution compared to lane 1. Theoretically you would think that sorting would improve barcode detection. 

### Subset data based on QC metrics

```{r}
dataQC <- lapply(dataSeurat, function(x) subset(x, subset = (nFeature_RNA >= qcValues[1]) & (nCount_RNA >=qcValues[2]) & (percent.mt < qcValues[3])))
```

## [Add CR CRISPR Calls]{.underline}

Here, we merge the oBC calls with the dataset and 

```{r}
# import crispr calls for each lane
# KEEP ORDER OF LANES THE SAME AS SEURAT OBJECT
crisprCalls <- vector(mode = "list")
crisprCalls[["lane1"]] <- read_csv("C:/Users/Eric/My Drive/Lab/datasets/EYW/SIG04_10x_240816/cellranger_outs/SIG04_lane1_GEX_oBC/crispr_analysis/protospacer_calls_per_cell.csv")
crisprCalls[["lane2"]] <- read.csv("C:/Users/Eric/My Drive/Lab/datasets/EYW/SIG04_10x_240816/cellranger_outs/SIG04_lane2_GEX_oBC/crispr_analysis/protospacer_calls_per_cell.csv")

# replace feature_call with names
featureTib <- read_csv("C:/Users/Eric/My Drive/Lab/datasets/EYW/SIG04_10x_240816/cellranger_inputs/feature_ref_SIG04.csv") %>%
  select(id,name)
for(i in 1:length(crisprCalls)){
  df <- crisprCalls[[i]] %>%
    separate_rows(feature_call, sep = "\\|") %>%
    left_join(featureTib, by = c("feature_call" = "id")) %>%
    group_by(group = cell_barcode) %>%
    summarize(name = str_c(name, collapse = "|")) %>%
    ungroup()
  
  crisprCalls[[i]]$feature_call <- df$name
}

for(i in 1:length(dataQC)){
  # get name of sample
  sample <- names(dataQC)[i]
  
  # extract metadata
  metadata <- dataQC[[sample]]@meta.data %>%
    as_tibble(rownames = "cell_barcode")
  
  # match metadata with crispr calls
  metadata <- left_join(metadata, crisprCalls[[sample]]) %>%
    mutate(oBC_classification = case_when(num_features == 1 ~ "singlet",
                                             num_features >= 2 ~ "multiplet",
                                             is.na(num_features) ~ "none"))
  metadataDF <- as.data.frame(metadata[,-1])
  rownames(metadataDF) <- metadata$cell_barcode
  
  # replace seurat metadata
  dataQC[[sample]]@meta.data <- metadataDF
}
```

```{r}
table(dataQC[[1]]$oBC_classification)
table(dataQC[[2]]$oBC_classification)
```

```{r}
table(dataQC[[1]]$oBC_classification)/ncol(dataQC[[1]])
table(dataQC[[2]]$oBC_classification)/ncol(dataQC[[2]])
```

```{r}
# Calculate a UMAP embedding of the HTO data
DefaultAssay(dataQC$lane1) <- "oBC"
DefaultAssay(dataQC$lane2) <- "oBC"

dataQC <- lapply(dataQC, function(x) NormalizeData(x, assay="oBC",normalization.method = "CLR", verbose=F)) %>%
  lapply(function(x) ScaleData(x, features = rownames(x), verbose = FALSE)) %>%
  lapply(function(x) RunPCA(x, features = rownames(x),approx = F, npcs = nrow(x), verbose = F)) %>%
  lapply(function(x) RunUMAP(x, dims = 1:nrow(x), verbose = F))
```

```{r, fig.height=10, fig.width=10}
plotHTO <- function(x){
  Idents(x) <- "oBC_classification"
  # compare UMI number for singlets, doublets, negatives
  p2 <- VlnPlot(x, features = "nCount_oBC", pt.size = 0.1, log = TRUE, alpha = 0.5)
  # plot HTO data on tSNE
  p3 <- DimPlot(x)
  
  p2+p3
}

plotHTO(dataQC$lane1)
plotHTO(dataQC$lane2)
```

## [Merge 10x lanes into single dataset]{.underline}

```{r}
data <- merge(dataQC[[1]],dataQC[[2]],
              add.cell.ids = c("lane1","lane2"), project = "SIG04") %>%
  JoinLayers()

# rename orig.ident based on
data$orig.ident <- paste(data$orig.ident, gsub("_.*","", colnames(data)), sep = "-")

rm(dataQC)
```

## [Remove HTO negatives]{.underline}

```{r}
table(data$oBC_classification)

table(data$oBC_classification)/ncol(data)*100
```


```{r}
Idents(data) <- "oBC_classification"
dataSub <- subset(data, idents = c("singlet","none"))
```

```{r}
dataSub@meta.data %>%
  group_by(oBC_classification) %>%
  summarize(num_cells = n()) %>%
  mutate(percent = num_cells/sum(num_cells)*100)
```

## [Score for sources of variation]{.underline}

### Cell cycle variation
```{r}
DefaultAssay(dataSub) <- "RNA"
dataSub <- NormalizeData(dataSub)
# Identify the most variable genes
dataSub <- FindVariableFeatures(dataSub, 
                     selection.method = "vst",
                     nfeatures = 2000, 
                     verbose = FALSE)
# Scale the counts
dataSub <- ScaleData(dataSub)

cellCycleGenes <- read_csv("C:/Users/Eric/My Drive/Lab/datasets/gene_sets/cell_cycle_genes_seurat_mouse.csv")
# cellCycleGenes <- read_csv("/Users/wange7/Library/CloudStorage/GoogleDrive-ericwang314@gmail.com/My Drive/Lab/datasets/gene_sets/cell_cycle_genes_seurat_mouse.csv")

dataSub <- CellCycleScoring(dataSub, s.features = cellCycleGenes$s_genes,
                            g2m.features = cellCycleGenes$g2m_genes)
```

```{r, fig.width=8, fig.height=4}
# Plot the PCA colored by cell cycle phase
dataSub <- RunPCA(dataSub, verbose = FALSE)
DimPlot(dataSub,
        reduction = "pca",
        group.by= "Phase",
        split.by = "Phase")
```

### Mitochondrial variation

```{r, fig.width = 12, fig.height=4}
# Check quartile values
mitoQuart <- summary(dataSub@meta.data$mitoRatio)

# Turn mitoRatio into categorical factor vector based on quartile values
dataSub@meta.data$mitoFr <- cut(dataSub@meta.data$mitoRatio, 
                   breaks=c(-Inf, mitoQuart[2], mitoQuart[3], mitoQuart[5], Inf), 
                   labels=c("Low","Medium","Medium high", "High"))

DimPlot(dataSub,
        reduction = "pca",
        group.by= "mitoFr",
        split.by = "mitoFr")
```

## [Data normalization/scaling]{.underline}

### SCTransform

```{r}
options(future.globals.maxSize= 10000*1024^2)
# perform SCT normalization and find variable features
dataSub <- SCTransform(dataSub, vst.flavor = "v2",
                       vars.to.regress = c("orig.ident"),
                       variable.features.n = 3100)

# remove TCR genes from variable features
featureVar <- VariableFeatures(dataSub, assay = "SCT")
TCRgenes <- grep("^Tra[vj]|^Trb[vdj]",featureVar)
BCRgenes <- grep("^Igk[vc]|^Igl[vc]|^Ighv",featureVar)
viralGenes <- grep("^p139",featureVar)
featureVar[TCRgenes]
featureVar[BCRgenes]
featureVar[viralGenes]

VariableFeatures(dataSub, assay = "SCT") <- featureVar[-c(TCRgenes,BCRgenes,viralGenes)]

# scale SCT variable features in RNA assay
DefaultAssay(dataSub) <- "RNA"
dataSub <- ScaleData(dataSub, features = featureVar[-c(TCRgenes,BCRgenes,viralGenes)])
DefaultAssay(dataSub) <- "SCT"
```

### Normalize oBC

```{r}
dataSub <- NormalizeData(dataSub, assay = "oBC", normalization.method = "CLR")
```

## [Dim Reduction]{.underline}

```{r}
# run PCA using TCR/BCR excluded Variable Features
dataSub <- RunPCA(dataSub, npcs = 60, verbose = FALSE)
ElbowPlot(dataSub, ndims = 60) +
  ggtitle("dataSub Elbow") +
  scale_x_continuous(breaks = seq(0,60,3))
```

```{r}
dimValues <- 50
# run umap based on dimensionality from elbow plot
dataSub <- RunUMAP(dataSub, dims = 1:dimValues, verbose = F) %>%
  FindNeighbors(dims = 1:dimValues, verbose = FALSE) %>%
  FindClusters(resolution = seq(0,1,0.1), verbose = F,
               algorithm = 1)
```

```{r, fig.height=10, fig.width=6}
clustree(dataSub, prefix = "SCT_snn_res.") +
  ggtitle("dataSub Clustree")
```

```{r, fig.width=8, fig.height=15}
dataSub$seurat_clusters <- dataSub$SCT_snn_res.1
Idents(dataSub) <- dataSub$seurat_clusters
p1 <- DimPlot(dataSub) + theme(aspect.ratio = 1)
p2 <- DimPlot(dataSub, split.by = "orig.ident") + theme(aspect.ratio = 1)
p3 <- DimPlot(dataSub, split.by = "oBC_classification") + theme(aspect.ratio = 1)
p1/p3/p2
```

## [Save RDS]{.underline}

```{r}
saveRDS(dataSub, "C:/Users/Eric/My Drive/Lab/datasets/EYW/SIG04_10x_240816/seurat_outs/SIG04_singlet_negative.rds")
```
